### PR TITLE
Finetuning baseline

### PR DESCRIPTION
**Pull request by ejnnr**
This is a very simple detector: it finetunes the model on the clean reference dataset, and then assigns score by checking how different the logits of the finetuned and the original model are. (Higher logit difference = more anomalous)

Even without any hyperparameter tuning etc., this is a decent baseline, e.g. it get ~0.8 AUC for a WaNet/CIFAR10 model for which the Mahalanobis baseline appears to be worse than random (?! worth looking into)

Usage is e.g.:
```bash
python -m cupbearer.scripts.train_detector --detector finetuning --dir.full <output_dir> --task backdoor --task.backdoor wanet --task.path <dir_to_classifier_run> --detector.train.num_epochs 1
```

I upgraded the minimum Jax version to 0.4.16 because the new code needs the KL divergence in `jax.scipy.special`. Also had to make some small changes because the finetuning detector doesn't support layerwise scores, which some of the eval code assumed. The detector can probably be improved (e.g. it's unclear whether the current KL divergence is a good metric, and I didn't optimize the hyperparameters at all).

### CODE REVIEW

File: src/cupbearer/detectors/finetuning.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    ):
+        trainer_instance = FinetuningTrainer(
+            model=self.model,
+            loss_fn=self.loss_fn,
+            optimizer=optimizer.build(),
+            log_dir=self.save_path,
+            override_variables={
+                "params": self.params
+            },  # Use existing parameters for initialization
+            example_input=clean_dataset[0][0][None],
```
Comment (VRehnberg) [inline src/cupbearer/detectors/finetuning.py:61]: In case of image classification `clean_dataset[0][0][None]` is first image and add an extra last dimension (channel dimension?). It runs so it's probably correct, but I would have expected `clean_dataset[0][0][None, ...]` to be what we wanted?

File: src/cupbearer/detectors/finetuning.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        finetuned_p = jax.nn.softmax(finetuned_output, axis=-1)
+        original_p = jax.nn.softmax(original_output, axis=-1)
+
+        # Check whether we're going to get infinities:
+        if jnp.any(jnp.logical_and(finetuned_p == 0, original_p > 0)):
+            # We'd get an error anyway once we compute eval metrics, better to give
+            # a more specific one here.
+            raise ValueError("Infinite KL divergence")
+
+        # TODO: which direction do we want here? Or maybe just some other metric?
```
Comment (VRehnberg) [inline src/cupbearer/detectors/finetuning.py:109]: I'd say we should consider the finetuned distribution to be the "ground truth" as we're comparing to see if we're surprised by the original output. So this would be D(finetuned_p||original_p) which (checking rel_entr docs) would be the other way around.

I'm unsure about this so please convince me otherwise.

File: src/cupbearer/scripts/train_classifier.py [hunk 3]
```diff
@@ -22,8 +22,9 @@ def create_functions(self):
         def losses(params, state, batch):
             images, labels = batch
             logits = state.apply_fn({"params": params}, images)
-            one_hot = jax.nn.one_hot(labels, self.num_classes)
-            losses = optax.softmax_cross_entropy(logits=logits, labels=one_hot)
+            losses = optax.softmax_cross_entropy_with_integer_labels(
```
Comment (VRehnberg) [inline src/cupbearer/scripts/train_classifier.py:25]: Nice

**Review by VRehnberg — COMMENTED** [review]
Looks good only some minor notes.

File: pyproject.toml [hunk 4]
```diff
@@ -17,7 +17,7 @@ classifiers = [
 requires-python = ">=3.10"
 dependencies = [
     "flax~=0.7.0",
-    "jax~=0.4.13",
+    "jax~=0.4.16",
```
[the remark this answers is not in the document]
Comment (ejnnr) [reply pyproject.toml:20]: Hm, I didn't get a deprecation warning weirdly, thanks for mentioning that! I've pushed a change that fixes the type annotations. Out of curiosity, are there any other warnings you get?

File: src/cupbearer/detectors/anomaly_detector.py [hunk 5]
```diff
@@ -133,13 +133,15 @@ def eval(
...
         for batch in test_loader:
             inputs, new_labels = batch
-            new_scores = self.layerwise_scores(inputs)
-            scores.append(new_scores.mean(axis=0))
-            # Sum over batch axis
-            layer_scores = layer_scores + new_scores.sum(axis=1)
-            num_elements += new_scores.shape[1]
+            try:
+                new_scores = self.layerwise_scores(inputs)
+                scores.append(new_scores.mean(axis=0))
```
[the remark this answers is not in the document]
Comment (ejnnr) [reply src/cupbearer/detectors/anomaly_detector.py:138]: I don't think I follow---the intended behavior is that if `scores()` is overriden and layerwise_scores rasises a NotImplementedError, we use the new `scores` method in the child class. Are you saying that isn't what will happen, or that there's something wrong with that?

If we add a final decorator inside `AnomalyDetector`, that would currently break the new finetuning detector since it overrides `scores()` (because it can't implement scoring layerwise). But maybe I'm misunderstanding your suggestion?

File: src/cupbearer/detectors/finetuning.py (hunk 1 above)
Comment (ejnnr) [reply src/cupbearer/detectors/finetuning.py:61]: `[None]` is equivalent to `[None, ...]`, it adds a singleton dimension at the front, not the back. That's the batch dimension (which we removed by indexing into the first batch before). But I agree your version is more readable and I should also add a comment explaining all the indices here.

File: src/cupbearer/detectors/finetuning.py (hunk 2 above)
Comment (ejnnr) [reply src/cupbearer/detectors/finetuning.py:111]: Hm, I feel like this is a bit unprincipled anyway (it's not clear why there should be any asymmetry). So that points towards some other metric. But I checked some previous work by Redwood and they're using KL divergence with D(original || finetuned) apparently. I think for now let's just stick with that, I'll update the comment.

Empirically, it seems that everything works similarly well on a very quick test (including weird stuff like an L2 distance between probability distributions). If we think the metric is important, we should actually carefully compare different options at some point, but that's pretty far down my list of things to do for now. Might be more appealing if this baseline looks to be the strongest one after a while.

File: pyproject.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (ejnnr) [reply pyproject.toml]: I actually ended up downgrading the requirement to 0.4.11 since that's the latest version supported by jax-metal (which is a lot faster at least for me on an M1 chip)

[the remark this answers is not in the document]
Comment (VRehnberg) [reply pyproject.toml]: There is deprecation warning on jax.core.Shape as well (didn't see it at first), this is with using 0.4.18 and can't see this mentioned Jax's changelog. All warnings:
```
=============================== warnings summary ===============================
../cupbearer-env/lib/python3.10/site-packages/flax/linen/module.py:77
  /mimer/NOBACKUP/groups/ml-safety/vikren/mad/cupbearer-env/lib/python3.10/site-packages/flax/linen/module.py:77: DeprecationWarning: jax.random.KeyArray is deprecated. Use jax.Array for ann
otations, and jax.dtypes.issubdtype(arr.dtype, jax.dtypes.prng_key) for runtime detection of typed prng keys.
    KeyArray = Union[jax.Array, jax.random.KeyArray]  # pylint: disable=invalid-name

../cupbearer-env/lib/python3.10/site-packages/flax/linen/recurrent.py:45
  /mimer/NOBACKUP/groups/ml-safety/vikren/mad/cupbearer-env/lib/python3.10/site-packages/flax/linen/recurrent.py:45: DeprecationWarning: jax.random.KeyArray is deprecated. Use jax.Array for 
annotations, and jax.dtypes.issubdtype(arr.dtype, jax.dtypes.prng_key) for runtime detection of typed prng keys.
    PRNGKey = jax.random.KeyArray

../cupbearer-env/lib/python3.10/site-packages/flax/linen/recurrent.py:753
  /mimer/NOBACKUP/groups/ml-safety/vikren/mad/cupbearer-env/lib/python3.10/site-packages/flax/linen/recurrent.py:753: DeprecationWarning: jax.random.KeyArray is deprecated. Use jax.Array for
 annotations, and jax.dtypes.issubdtype(arr.dtype, jax.dtypes.prng_key) for runtime detection of typed prng keys.
    init_key: Optional[random.KeyArray] = None,

../cupbearer-env/lib/python3.10/site-packages/flax/linen/recurrent.py:979
  /mimer/NOBACKUP/groups/ml-safety/vikren/mad/cupbearer-env/lib/python3.10/site-packages/flax/linen/recurrent.py:979: DeprecationWarning: jax.random.KeyArray is deprecated. Use jax.Array for
 annotations, and jax.dtypes.issubdtype(arr.dtype, jax.dtypes.prng_key) for runtime detection of typed prng keys.
    init_key: Optional[random.KeyArray] = None,

../cupbearer-env/lib/python3.10/site-packages/flax/linen/recurrent.py:1003
  /mimer/NOBACKUP/groups/ml-safety/vikren/mad/cupbearer-env/lib/python3.10/site-packages/flax/linen/recurrent.py:1003: DeprecationWarning: jax.random.KeyArray is deprecated. Use jax.Array fo
r annotations, and jax.dtypes.issubdtype(arr.dtype, jax.dtypes.prng_key) for runtime detection of typed prng keys.
    init_key: Optional[random.KeyArray] = None,

../cupbearer-env/lib/python3.10/site-packages/flax/linen/stochastic.py:28
  /mimer/NOBACKUP/groups/ml-safety/vikren/mad/cupbearer-env/lib/python3.10/site-packages/flax/linen/stochastic.py:28: DeprecationWarning: jax.random.KeyArray is deprecated. Use jax.Array for annotations, and jax.dtypes.issubdtype(arr.dtype, jax.dtypes.prng_key) for runtime detection of typed prng keys.
    KeyArray = Union[jax.Array, jax.random.KeyArray]

../cupbearer-env/lib/python3.10/site-packages/chex/_src/pytypes.py:53
  /mimer/NOBACKUP/groups/ml-safety/vikren/mad/cupbearer-env/lib/python3.10/site-packages/chex/_src/pytypes.py:53: DeprecationWarning: jax.core.Shape is deprecated. Use Shape = Sequence[int | Any].
    Shape = jax.core.Shape

../cupbearer-env/lib/python3.10/site-packages/cupbearer/models/computations.py:239
  /mimer/NOBACKUP/groups/ml-safety/vikren/mad/cupbearer-env/lib/python3.10/site-packages/cupbearer/models/computations.py:239: DeprecationWarning: jax.random.KeyArray is deprecated. Use jax.Array for annotations, and jax.dtypes.issubdtype(arr.dtype, jax.dtypes.prng_key) for runtime detection of typed prng keys.
    key: jax.random.KeyArray, shape: Sequence[int], dtype=jnp.float_, value=0.0

../cupbearer-env/lib/python3.10/site-packages/cupbearer/models/computations.py:245
  /mimer/NOBACKUP/groups/ml-safety/vikren/mad/cupbearer-env/lib/python3.10/site-packages/cupbearer/models/computations.py:245: DeprecationWarning: jax.random.KeyArray is deprecated. Use jax.Array for annotations, and jax.dtypes.issubdtype(arr.dtype, jax.dtypes.prng_key) for runtime detection of typed prng keys.
    key: jax.random.KeyArray, shape: Sequence[int], dtype=jnp.float_

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
```

File: src/cupbearer/detectors/anomaly_detector.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (VRehnberg) [reply src/cupbearer/detectors/anomaly_detector.py:138]: In cases where `scores()` is overridden and `layerwise_scores()` does not raise `NotImplementedError` you might still expect scores to be calculated by `scores()` instead of `layerwise_scores(inputs).mean(axis=0)`.

But, yes I didn't think this through.

We could still add check to see if scores are overridden (something like `type(self).scores == AnomalyDetector.scores`) but that is a lot less clean, so we could just skip this. Though, a comment that `scores()` are unused if `layerwise_scores()` is implemented would be nice.

File: pyproject.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (ejnnr) [reply pyproject.toml]: Thanks, not sure why I'm not seeing these. I generally dislike the many warnings a lot of ML code produces and would like to avoid them if possible. Seems like I do see warnings in the pytorch refactor though, so probably won't look into this (and just fix the pytorch ones where possible)

File: src/cupbearer/detectors/anomaly_detector.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ejnnr) [reply src/cupbearer/detectors/anomaly_detector.py:138]: Oh, I see what you mean. Yeah, overriding `scores` while also implementing `layerwise_scores` is something I didn't intend to support atm. I actually think your suggested check is best here---it does feel a bit strange, but that's probably the fault of the overall design. If we have a check at least things won't fail quietly.

**Review by VRehnberg — COMMENTED** [review]
Lgtm
